### PR TITLE
Pull generate library up

### PR DESCRIPTION
**Pull request by ledogar**
### Description of your changes
Refactors the generateLibrary() method so that it can be pulled up to the parent Lipid class.
Two of the lipid subclasses override this method:
- LysoCL uses a modified `chains` variable
- MIP2C attempts to use a modified `chains` variable, but appears to have a bug (not addressed in this PR)

### Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Refactor (no change to functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation add / update

### (If applicable) How has this been tested?
Passes test suite

### CODE REVIEW

**Comment by ledogar** [comment]
@PMSeitzer I think I found a bug in MIP2C.py.  If you look in the master branch, generateLibrary is defined twice.  The second definition shadows the first.  When I tried to unshadow the first, I encountered a runtime error in how the `chains` variable was being customized.  See context comment

File: calicolipidlibrary/MIP2C.py [hunk 1]
```diff
@@ -45,39 +45,23 @@ def theoreticalDigest(self):
 			return(FRAGMENTS)
 
 
-		@classmethod
-		def generateLibrary(cls,target=None, mode="pos"):
+
+		def generateLibrary(self, target=None, mode="pos"):
```
Comment (ledogar) [inline calicolipidlibrary/MIP2C.py:49]: This is the first definition of generateLibrary().  It has a customized `chains` variable

File: calicolipidlibrary/MIP2C.py [hunk 2]
```diff
@@ -45,39 +45,23 @@ def theoreticalDigest(self):
...
 			if target: handle = open(target, 'a+')
-			if mode == "pos":  
+			if mode == "pos":
 				adduct_set = self.pos_adduct_set
 			elif mode =="neg":
 				adduct_set = self.neg_adduct_set
 			for c1 in self.chain1_ranges:
 				for c2 in self.chain2_ranges:
 					for adduct in adduct_set:
-						x  = MIP2C("MIP2C",[c1[0],c2[0]], [c1[1],c2[1]], [c1[2],c2[2]],  adduct=adduct)
```
Comment (ledogar) [inline calicolipidlibrary/MIP2C.py:58]: This is in the first definition of generateLibrary().  This is the customized `chains` variable.  It appears to have a syntax error: the MIP2C constructor expects `(name, chainsList, adduct)` but receives `(name, chain, chain, chain, adduct)` possibly related to in-line square-bracket nesting

File: calicolipidlibrary/MIP2C.py [hunk 3]
```diff
@@ -45,39 +45,23 @@ def theoreticalDigest(self):
...
-						x  = MIP2C("MIP2C",[c1[0],c2[0]], [c1[1],c2[1]], [c1[2],c2[2]],  adduct=adduct)						
-						content = x.printNist()
-						if target: handle.write(content)
-						else: sys.stdout.write(content)
-			if target: handle.close()
-
-
-
-		@classmethod
-		def generateLibrary(cls,target=None, mode="pos"):
```
Comment (ledogar) [inline calicolipidlibrary/MIP2C.py:67]: This is the second definition of generateLibrary() which shadows the first, hiding its customized chains behavior (which has a syntax error)

File: calicolipidlibrary/MIP2C.py [hunk 4]
```diff
@@ -45,40 +45,6 @@ def theoreticalDigest(self):
 			return(FRAGMENTS)
 
 
-		@classmethod
-		def generateLibrary(cls,target=None, mode="pos"):
```
Comment (ledogar) [inline calicolipidlibrary/MIP2C.py:49]: This was the first definition of generateLibrary().  It defined a customized `chains` variable.

File: calicolipidlibrary/MIP2C.py [hunk 5]
```diff
@@ -45,40 +45,6 @@ def theoreticalDigest(self):
...
-			if target: handle = open(target, 'a+')
-			if mode == "pos":  
-				adduct_set = cls.pos_adduct_set
-			elif mode =="neg":
-				adduct_set = cls.neg_adduct_set
-			parent = cls.__bases__[0]
-			class_name = cls.__name__ 
-			for c in parent.chain_sets:
-				for adduct in adduct_set:
-					x  = cls(class_name, c, adduct=adduct)
```
Comment (ledogar) [inline calicolipidlibrary/MIP2C.py:77]: Note the second generateLibrary definition removed the customized `chains` variable defined in the first generateLibrary definition.

File: calicolipidlibrary/MIP2C.py [hunk 6]
```diff
@@ -45,40 +45,6 @@ def theoreticalDigest(self):
...
-		def generateLibrary(cls,target=None, mode="pos"):
-			if target: handle = open(target, 'a+')
-			if mode == "pos":  
-				adduct_set = self.pos_adduct_set
-			elif mode =="neg":
-				adduct_set = self.neg_adduct_set
-			for c1 in self.chain1_ranges:
-				for c2 in self.chain2_ranges:
-					for adduct in adduct_set:
-						x  = MIP2C("MIP2C",[c1[0],c2[0]], [c1[1],c2[1]], [c1[2],c2[2]],  adduct=adduct)
```
Comment (ledogar) [inline calicolipidlibrary/MIP2C.py:58]: This line was in the first definition of generateLibrary(). It defined a customized chains variable. The definition appears to have a syntax error: the MIP2C constructor expects `(name, chainsList, adduct)` but receives `(name, chain, chain, chain, adduct)`, maybe a typo, possibly related to in-line square-bracket nesting.

File: calicolipidlibrary/MIP2C.py [hunk 7]
```diff
@@ -45,40 +45,6 @@ def theoreticalDigest(self):
...
-						x  = MIP2C("MIP2C",[c1[0],c2[0]], [c1[1],c2[1]], [c1[2],c2[2]],  adduct=adduct)						
-						content = x.printNist()
-						if target: handle.write(content)
-						else: sys.stdout.write(content)
-			if target: handle.close()
-
-
-
-		@classmethod
-		def generateLibrary(cls,target=None, mode="pos"):
```
Comment (ledogar) [inline calicolipidlibrary/MIP2C.py:67]: This was the second definition of generateLibrary(), which shadowed the first.

**Comment by ledogar** [comment]
@PMSeitzer The MIP2C issue is kindof confusing in this PR.  The PR attempts to not change the code behavior for MIP2C.  However there's an error buried in the code which might be worth a separate discussion (outside of this PR)

**Review by PMSeitzer — COMMENTED** [review]
lgtm
Thanks for doing this refactoring!

**Pull request merged**